include/oneapi/dpl/pstl/utils.h

```diff
@@ -126,6 +128,32 @@ class __pstl_assign
     }
 };
 
+template <typename _Proj>
+struct __projection_deref
+{
+    //'mutable' is to relax the requirements for a user comparator or/and projection type operator() may be non-const
+    mutable _Proj __proj;
+
+    template <typename _Iterator>
+    decltype(auto)
+    operator()(_Iterator __it) const
+    {
+        if constexpr (std::is_lvalue_reference_v<decltype(*__it)>)
+        {
+            return std::invoke(__proj, *__it);
+        }
+        else
+        {
+            using __proj_result_t = decltype(std::invoke(__proj, *__it));
+            using __proj_result_decayed_t = std::decay_t<__proj_result_t>;
+
+            // create temporary copy of the projection and return it as prvalue
+            // - for that we should create and return temporary object (but not local variable)
+            return __proj_result_decayed_t{std::invoke(__proj, *__it)};
```

Comment:
Is it guaranteed that the type is copyable?

Reply:
Its a good question.  Perhaps it is best just to use `std::invoke(__proj1,*it)` inline everywhere instead of `__proj1_deref(it)` after the declaration of `oneapi::dpl::__internal::__projection_deref<_Proj1> __proj1_deref{__proj1};`.  

All of the complexity of the lambda or struct including dealing with this question of copyability of the projected result is to try to handle different return type value categories.  If we just do it inline, I think there should be no problems in this regard, and the complexity of the code isn't that much higher. Perhaps any complexity we save in consolidation here is created by its existence in the first place.

Reply:
As far as I understand, it will work in the case if copy disabled too.
Example: https://godbolt.org/z/jsfoq9YsM
Justification: https://en.cppreference.com/w/cpp/language/copy_elision.html :
```
When a class object target is copy-initialized with a temporary class object obj that has not been bound to a reference, the copy-initialization can be omitted by constructing obj directly into target. This variant of copy elision is known as unnamed return value optimization (URVO). Since C++17, URVO is mandatory and no longer considered a form of copy elision; see below.
```

Reply:
Alternative approach prepared in separate PR https://github.com/uxlfoundation/oneDPL/pull/2412

Reply:
Thanks for the investigation Sergey.  I think you are correct that this works with copy disabled.  That being said, I think I prefer inlining `std::invoke` with the explicit dereference.  It lessens our need to consider these complicated topics, and I think the lines with `std::invoke` and explicit dereference are not much more complicated ultimately.